src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSpec.scala

```diff
@@ -0,0 +1,314 @@
+    } thenReturn Future.successful(Some("1.2.3.4"))
+
+    createClusterSupervisor(dao) ! ClusterCreated(creatingCluster)
+
+    expectMsgClass(1 second, classOf[Terminated])
+    val updatedCluster = dbFutureValue { _.clusterQuery.getByName(creatingCluster.googleProject, creatingCluster.clusterName) }
+    updatedCluster shouldBe 'defined
+    updatedCluster.map(_.status) shouldBe Some(ClusterStatus.Running)
+    updatedCluster.flatMap(_.hostIp) shouldBe Some("1.2.3.4")
+  }
+
+  // Pre:
+  // - cluster exists in the DB with status Creating
+  // - dataproc DAO returns status CREATING, UNKNOWN, or UPDATING
+  // Post:
+  // - cluster is not changed in the DB
+  // - monitor actor does not shut down
+  Seq(ClusterStatus.Creating, ClusterStatus.Updating, ClusterStatus.Unknown).map { status =>
+    it should s"monitor $status status" in isolatedDbTest {
+      dbFutureValue { _.clusterQuery.save(creatingCluster) } shouldEqual creatingCluster
+
+      val dao = mock[DataprocDAO]
+      when {
+        dao.getClusterStatus(eqq(creatingCluster.googleProject), eqq(creatingCluster.clusterName))(any[ExecutionContext])
+      } thenReturn Future.successful(status)
+
+      createClusterSupervisor(dao) ! ClusterCreated(creatingCluster)
+
+      expectNoMsg(1 second)
+
+      val updatedCluster = dbFutureValue { _.clusterQuery.getByName(creatingCluster.googleProject, creatingCluster.clusterName) }
+      updatedCluster shouldBe 'defined
+      updatedCluster shouldBe Some(creatingCluster)
+    }
+  }
+
+  // Pre:
+  // - cluster exists in the DB with status Creating
+  // - dataproc DAO returns status RUNNING, but no IP address
+  // Post:
+  // - cluster is not changed in the DB
+  // - monitor actor does not shut down
+  it should "keep monitoring in RUNNING state with no IP" in isolatedDbTest {
+    dbFutureValue { _.clusterQuery.save(creatingCluster) } shouldEqual creatingCluster
+
+    val dao = mock[DataprocDAO]
+    when {
+      dao.getClusterStatus(eqq(creatingCluster.googleProject), eqq(creatingCluster.clusterName))(any[ExecutionContext])
+    } thenReturn Future.successful(ClusterStatus.Running)
+
+    when {
+      dao.getClusterMasterInstanceIp(eqq(creatingCluster.googleProject), eqq(creatingCluster.clusterName))(any[ExecutionContext])
+    } thenReturn Future.successful(None)
+
+    createClusterSupervisor(dao) ! ClusterCreated(creatingCluster)
+
+    expectNoMsg(1 second)
+
+    val updatedCluster = dbFutureValue { _.clusterQuery.getByName(creatingCluster.googleProject, creatingCluster.clusterName) }
+    updatedCluster shouldBe 'defined
+    updatedCluster shouldBe Some(creatingCluster)
+  }
+
+  // Pre:
+  // - cluster exists in the DB with status Creating
+  // - dataproc DAO returns status ERROR, but no error code
+  // Post:
+  // - cluster is not changed in the DB
+  // - monitor actor does not shut down
+  it should "keep monitoring in ERROR state with no error code" in isolatedDbTest {
```

Comment:
Why does it behave differently with no error code?

Reply:
Currently, a Failed cluster requires an error code. The ClusterMonitorActor uses the error code to determine what to do next: either delete and recreate the cluster, or set the clusterStatus to ERROR and stop monitoring it. If there is no error code (or something else went wrong in getting the error details), we generate a "not ready" message and check again the next pass.

Reply:
ok in `checkCluster` / `getClusterErrorDetails` yes